### PR TITLE
Added ts-prune to detect unused code

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## What was changed
Adds [ts-prune](https://github.com/nadeesha/ts-prune), a nifty tool that detects unused code.

Copying contents & threads from https://github.com/JoshuaKGoldberg/sdk-typescript/pull/2 with the addition of a CI job.

## Why?
It found a few unused items and would be useful in CI. [[Slack thread](https://temporaltechnologies.slack.com/archives/C01FT8U10GK/p1637960205341800)]

## Checklist

1. How was this tested:
See CI job, once it's running.

Note that I can't run the GHA job as I'm an external contributor working from a fork. Someone will have to try that out for me / set me up to do it.

### CODE REVIEW

File: packages/worker/src/utils.ts [hunk 1]
```diff
@@ -1,8 +1,5 @@
-export const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));
-export const identity = <T>(x: T): T => x;
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/utils.ts:2]: `identity` was unused. `sleep` was only used in a test file so I moved it there.

File: packages/worker/src/utils.ts [hunk 2]
```diff
@@ -1,8 +1,5 @@
...
-
-export const GiB = 1024 ** 3;
 export const MiB = 1024 ** 2;
+export const GiB = MiB * 1024;
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/utils.ts:2]: `MiB` was unused but factors into `GiB` nicely.

File: tsconfig.prune.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  "files": [
+    "./packages/activity/src/index.ts",
+    "./packages/client/src/index.ts",
+    "./packages/common/src/index.ts",
+    "./packages/create-project/src/index.ts",
+    "./packages/meta/src/index.ts",
+    "./packages/worker/src/index.ts",
+    "./packages/worker/src/workflow/workflow-worker-thread.ts",
+    "./packages/workflow/src/index.ts"
+  ]
```
Comment (JoshuaKGoldberg) [inline tsconfig.prune.json:17]: These `files` act as the entry points for unused code detection. In other words, code things that are transitively imported by these files are counted as used.

File: packages/worker/src/rxutils.ts [hunk 4]
```diff
@@ -6,20 +6,9 @@ interface StateAndOptionalOutput<T, O> {
   output?: O;
 }
 
-export type StateAndOutput<T, O> = Required<StateAndOptionalOutput<T, O>>;
+export const definitelyUnusedRemoveOnceWeKnowCiWorks = true;
 
-export function mapWithState<T, I, O>(
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/rxutils.ts:11]: Totally unused 🤘

File: packages/worker/src/index.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@
 export { State, Worker, ActivityOptions, DataConverter, defaultDataConverter, RetryOptions, errors } from './worker';
 export { WorkerOptions, CompiledWorkerOptions } from './worker-options';
 export { ServerOptions, TLSConfig } from './server-options';
-export { Core } from './core';
+export { Core, CompiledCoreOptions, CoreOptions, TelemetryOptions } from './core';
```
Comment (JoshuaKGoldberg) [inline packages/worker/src/index.ts:14]: These options interfaces are used in `Core`'s functions but not exported. Perhaps not required for unused code detection, but a nice to export.

File: package.json [hunk 6]
```diff
@@ -45,6 +46,7 @@
     "ci-nightly-sampler": "node ./packages/test/lib/load/all-in-one.js --server-address $TEMPORAL_TESTING_SERVER_URL --otel-url='' --iterations 1000 --min-wfs-per-sec 5 --max-cached-wfs 500 --workflow yummy-sampler-mode",
     "wait-namespace": "node ./scripts/wait-on-temporal.mjs",
     "lint": "eslint packages/*/src --ext .ts --no-error-on-unmatched-pattern && prettier --end-of-line auto --check .",
+    "lint.prune": "! ts-prune -p tsconfig.prune.json --ignore \"used in module\" | grep -v \".d.ts\"",
```
Comment (JoshuaKGoldberg) [inline package.json:49]: Until https://github.com/nadeesha/ts-prune/issues/156 is resolved, `--skip` doesn't work. Hence the `| grep -v`. 

By default `grep` returns `0` if found and `1` or not. We actually want the opposite... so the `!` at the beginning inverts the exist code. 🧠

File: tsconfig.prune.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  "files": [
+    "./packages/activity/src/index.ts",
+    "./packages/client/src/index.ts",
+    "./packages/common/src/index.ts",
+    "./packages/create-project/src/index.ts",
+    "./packages/meta/src/index.ts",
+    "./packages/worker/src/index.ts",
+    "./packages/worker/src/workflow/workflow-worker-thread.ts",
+    "./packages/workflow/src/index.ts"
```
Comment (bergundy) [inline tsconfig.prune.json:16]: We should add `./packages/workflow/src/worker-interface.ts` too

File: package.json (hunk 6 above)
Comment (bergundy) [reply package.json:49]: Will this syntax work on windows?
Might be worth adding the prune CI step to the build-lint-test CI job to verify.

File: packages/worker/src/index.ts (hunk 5 above)
Comment (bergundy) [reply packages/worker/src/index.ts:14]: 👍

File: package.json (hunk 6 above)
Comment (JoshuaKGoldberg) [reply package.json:49]: Alas, it will not -- Windows also doesn't have `!` or `grep` 😞....

Comment (JoshuaKGoldberg) [reply package.json:49]: Aha, the upstream issue's fixed was released!

File: tsconfig.prune.json (hunk 7 above)
Comment (bergundy) [reply tsconfig.prune.json:16]: @JoshuaKGoldberg I want to merge this, can you please add the missing file?

Comment (JoshuaKGoldberg) [reply tsconfig.prune.json:16]: Ah, sorry, I'd missed this notification! Merging now.

**Review by bergundy — APPROVED** [review]

**Pull request merged**